### PR TITLE
feat(internal): Replace compression with faster alternative

### PR DESCRIPTION
**Pull request by zekena2**
- [X] Updated associated README.md.
- [X] Wrote appropriate unit tests.
- [X] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

This PR uses the klauspost/compress and pgzip library instead of the standard library as it's more performant as disscused [here](https://github.com/influxdata/telegraf/pull/13245#discussion_r1194135580) and adds zstd as an option using the same library. Moreover, it adds a possibility to decide the encoding level. Benchmark tests could be added later.

helps https://github.com/influxdata/telegraf/pull/13245

Some benchmarks have been done to compare the different libraries used, the detailed comparisons is in the PR comments. This benchmark is done running `go test -bench . -v`

goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/internal
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

| compression            | stdlib           |    this PR.     | speedup |
|------------------------|------------------|----------------------|---|
| gzip encode            |     21,172 ns/op |      **4,072 ns/op**    | ~5x |
| gzip decode            |      1,478 ns/op |       1,216 ns/op    | - |
| gzip encode/decode     |     24,670 ns/op |      **5,160 ns/op**    | ~5x  
| gzip big encode        | 60,065,339 ns/op |  **3,321,840 ns/op** | ~20x |
| gzip big decode        |  5,561,827 ns/op |    4,830,250 ns/op   | - |
| gzip big encode/decode | 65,026,887 ns/op | **8,077,422 ns/op** | ~8x |
| zlip encode            |     22,385 ns/op |      **4,200 ns/op** | ~5x |
| zlip decode            |      6,369 ns/op |        5,975 ns/op   | - |
| zlip encode/decode     |     41,488 ns/op |     **13,148 ns/op** | ~3x |
| identity encode/decode |      616.4 ns/op |        719.6 ns/op   | - |

### CODE REVIEW

**Comment by zekena2** [comment]
@srebhan Please take a look I think I'll need some help with adding these benchmark tests.

File: internal/content_coding.go [hunk 1]
```diff
@@ -3,10 +3,11 @@ package internal
...
 	"bufio"
 	"bytes"
-	"compress/gzip"
-	"compress/zlib"
 	"errors"
 	"fmt"
+	"github.com/klauspost/compress/zlib"
+	"github.com/klauspost/compress/zstd"
+	gzip "github.com/klauspost/pgzip"
 	"io"
```
Comment (srebhan) [inline internal/content_coding.go:11]: We group import by _built-in_, _3rd party_ and _telegraf-internal_, each group sorted alphabetically and separated by a newline. So you should do
```suggestion
	"io"

	"github.com/klauspost/compress/zlib"
	"github.com/klauspost/compress/zstd"
	gzip "github.com/klauspost/pgzip"
```

File: internal/content_coding.go [hunk 2]
```diff
@@ -72,12 +73,14 @@ func (r *GzipReader) Read(b []byte) (int, error) {
 }
 
 // NewContentEncoder returns a ContentEncoder for the encoding type.
-func NewContentEncoder(encoding string) (ContentEncoder, error) {
+func NewContentEncoder(encoding string, level int) (ContentEncoder, error) {
```
Comment (srebhan) [inline internal/content_coding.go:76]: Not a must, just an idea: Should we adopt the [functional option pattern](https://golang.cafe/blog/golang-functional-options-pattern.html) here? This would allow to set a default level and keep the API stable to what we had before. At the same time it allows to extend the encoders later with more options...

File: internal/content_coding.go [hunk 3]
```diff
@@ -137,12 +142,13 @@ type GzipEncoder struct {
 	buf    *bytes.Buffer
 }
 
-func NewGzipEncoder() *GzipEncoder {
+func NewGzipEncoder(level int) (*GzipEncoder, error) {
 	var buf bytes.Buffer
+	w, err := gzip.NewWriterLevel(&buf, level)
```
Comment (srebhan) [inline internal/content_coding.go:147]: I would rather return a `nil` encoder to force the caller to check the error code
```suggestion
	w, err := gzip.NewWriterLevel(&buf, level)
	if err != nil {
		return nil, err
	}
```
You can then return `nil` for the error below.

Same for all other instances here.

File: internal/content_coding.go [hunk 4]
```diff
@@ -188,6 +195,21 @@ func (e *ZlibEncoder) Encode(data []byte) ([]byte, error) {
...
+	e, err := zstd.NewWriter(nil, zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(level)))
+	return &ZstdEncoder{
+		encoder: e,
+	}, err
+}
+
+func (e *ZstdEncoder) Encode(data []byte) ([]byte, error) {
+	return e.encoder.EncodeAll(data, make([]byte, 0, len(data))), nil
+}
+
```
Comment (srebhan) [inline internal/content_coding.go:212]: Could we split this into yet another PR. Sorry for being so fine-grained, but mixing topics in the PR usually leads to trouble when trying to pin-down errors introduced by some commits.

File: internal/content_coding_test.go [hunk 5]
```diff
@@ -113,3 +156,168 @@ func TestStreamGzipDecode(t *testing.T) {
...
+
+		require.Equal(b, "howdy", string(actual))
+
+		payload, err = enc.Encode([]byte("doody"))
+		require.NoError(b, err)
+
+		actual, err = dec.Decode(payload, maxDecompressionSize)
+		require.NoError(b, err)
+
+		require.Equal(b, "doody", string(actual))
```
Comment (srebhan) [inline internal/content_coding_test.go:198]: Just do the compression/decompression once and add a warmup before the benchmark to have the internal buffer allocated...
```suggestion
	data := []byte("This is a test message for compression and decompression benchmarking!")

	// Warm-up the decompression to not measure internal buffer allocations
	payload, err := enc.Encode(data)
	require.NoError(b, err)
	actual, err := dec.Decode(payload, maxDecompressionSize)
	require.NoError(b, err)
	require.Equal(b, data, actual)
	
	for n := 0; n < b.N; n++ {
		p, _ := enc.Encode(data)
		_, _ = dec.Decode(p, maxDecompressionSize)
```
You can do the same for the other algorithms below.

File: internal/content_coding_test.go [hunk 6]
```diff
@@ -113,3 +156,168 @@ func TestStreamGzipDecode(t *testing.T) {
...
+	}
+}
```
Comment (srebhan) [inline internal/content_coding_test.go:240]: You don't need this one.

File: internal/content_coding_test.go (hunk 5 above)
Comment (srebhan) [reply internal/content_coding_test.go:198]: If you want to, you can split the benchmark into one for compression and one for decompression, but for me having both in one is also ok.

File: internal/content_coding_test.go [hunk 7]
```diff
@@ -113,3 +156,168 @@ func TestStreamGzipDecode(t *testing.T) {
...
+	dec := NewZlibDecoder()
+
+	for n := 0; n < b.N; n++ {
+		payload, err := enc.Encode([]byte("howdy"))
+		require.NoError(b, err)
+
+		_, err = dec.Decode(payload, 3)
+		require.ErrorContains(b, err, "size of decoded data exceeds allowed size 3")
+	}
+}
```
Comment (srebhan) [inline internal/content_coding_test.go:270]: This should be a test instead.

**Review by srebhan — COMMENTED** [review]
Thanks for the split-out @zekena2! Overall the code looks good, some comments though... If we have the compression level as an (optional) "option" we could keep the interface. Furthermore, I added some comments regarding your benchmark. If you have questions, please feel free to ask!

File: internal/content_coding.go (hunk 2 above)
Comment (zekena2) [reply internal/content_coding.go:76]: I wanted something like that. I'll play with it more.

File: internal/content_coding_test.go (hunk 7 above)
Comment (zekena2) [reply internal/content_coding_test.go:270]: This already exists as a test. I'll just remove it.

File: internal/content_coding_test.go (hunk 6 above)
Comment (zekena2) [reply internal/content_coding_test.go:240]: Why though? Why having the ZstdReuse and removing this?

**Comment by zekena2** [comment]
How do we test these benchmarks? I don't see -bench when running `make test` for example and I don't see it in circle CI steps

File: internal/content_coding_test.go (hunk 6 above)
Comment (srebhan) [reply internal/content_coding_test.go]: You don't need it if you change the benchmark above as it will cover the reuse case already. :-) What I mean here is remove all the reuse tests and do it in the normal test. I don't think there is a point in measuring the allocation speed of the machine...

File: internal/content_coding_test.go [hunk 8]
```diff
@@ -113,3 +118,112 @@ func TestStreamGzipDecode(t *testing.T) {
...
+	for n := 0; n < b.N; n++ {
+		payload, err = enc.Encode([]byte("doody"))
+		require.NoError(b, err)
+
+		actual, err = dec.Decode(payload, maxDecompressionSize)
+		require.NoError(b, err)
+
+		require.Equal(b, "doody", string(actual))
+	}
+}
```
Comment (srebhan) [inline internal/content_coding_test.go:160]: I think I caused some confusion here... Sorry for that. For each algorithm, I want to see a benchmark for the encoding/decoding cycle. As the `for` loop is what determines the measurement you should trim that one down as much as possible. Therefore, please remove the equality check.

Furthermore, we should **only** check the "reuse" case as we otherwise only include the buffer allocation time which does not make sense IMO. Anyway, your "non-reuse" test won't work as `b.N` is usually large and after `b.N > 0` you reuse the buffer anyway as you do not recreate the encoder/decoders.

It would maybe be good to increase the message size. Therefore my suggestion is to have the following for all algorithms (`gzip` and `identity` IIRC):
```suggestion
func BenchmarkGzipEncodeDecode(b *testing.B) {
	data := []byte(strings.Repeat("-howdy stranger-", 64))

	enc, err := NewGzipEncoder(-1)
	require.NoError(b, err)
	dec := NewGzipDecoder()

	payload, err := enc.Encode(data)
	require.NoError(b, err)
	actual, err := dec.Decode(payload, maxDecompressionSize)
	require.NoError(b, err)
	require.Equal(b, data, actual)

	for n := 0; n < b.N; n++ {
		p, err := enc.Encode(data)
		require.NoError(b, err)

		_, err = dec.Decode(p, maxDecompressionSize)
		require.NoError(b, err)
	}
}
```

**Review by srebhan — COMMENTED** [review]
Some clarification regarding benchmarks (at least I hope so)... :-)

**Comment by srebhan** [comment]
@zekena2 the benchmark code looks good now. Looking forward to the "option" modification discussed above. If you find some time, you can try to run the benchmarks against the current and the new code to see how the speed improves...

**Comment by srebhan** [comment]
@zekena2 did you get a chance to look into the 'options'? I would really love to see this in 1.27...

**Comment by zekena2** [comment]
I didn't have time last couple of days but I'll try to get everything done before 1.27 release though. For the options feature I think if I didn't get it done by the weekend we should merge this PR and implement it in the future.

**Comment by srebhan** [comment]
Hmmm, is it OK for you if I work on the options part? I'm hesitant in switching all users to providing a param and then switch them back. ;-)

**Comment by zekena2** [comment]
yes you can work on it as well. What I was trying to do is similar to what the zstd package is doing with EOption and encoderOptions.
[here](https://github.com/klauspost/compress/blob/master/zstd/encoder_options.go)

**Comment by zekena2** [comment]
@srebhan Would you work on it? I don't think I'm able to make it as I wish it to be, it turns out to be easier without the options part.

**Review by powersj — APPROVED** [review]
@srebhan - assuming benchmarks show this as good, I am +1 on this

**Comment by srebhan** [comment]
For a 16MB input you get

| compression            | before           | new                   |
|------------------------|------------------|-----------------------|
| gzip big encode        | 37,801,517 ns/op |   **3,744,470 ns/op** |
| gzip big decode        |  4,568,003 ns/op |   **6,530,409 ns/op** |
| gzip big encode/decode | 43,386,038 ns/op |  **10,343,237 ns/op** |

which is a nice improvement... @powersj are you OK with dynamically switching depending on data-size using the proposed 1MB threshold?

**Comment by zekena2** [comment]
Is it often that data is written smaller than 1MB?

**Comment by powersj** [comment]
> Is it often that data is written smaller than 1MB?

heh I was about to say, how often is data written bigger than 1MB :P To cross that line you would need to be writing probably tens of thousands or so metrics at a time. I will admit this is entirely dependent on the size of the metric itself. 

However, the default batch size is 1000 and unless someone is modifying, which I know people do, I find it is unlikely that the majority of users are always writing >1MB.

> are you OK with dynamically switching depending on data-size using the proposed 1MB threshold?

My off the cuff thought is to say we should avoid this complexity and multiple paths, but a PR to look at might change my mind.

**Comment by srebhan** [comment]
I think complexity is not too high... Benchmark with the change:

| compression            | before           | new                  |
|------------------------|------------------|----------------------|
| gzip encode            |     16,622 ns/op |      16,329 ns/op    |
| gzip decode            |      1,590 ns/op |       1,635 ns/op    |
| gzip encode/decode     |     18,339 ns/op |      18,076 ns/op    |
| gzip big encode        | 37,801,517 ns/op |  **3,794,025 ns/op** |
| gzip big decode        |  4,568,003 ns/op |    4,469,810 ns/op   |
| gzip big encode/decode | 43,386,038 ns/op | **12,543,741 ns/op** |
| zlip encode            |     16,952 ns/op |      **3,348 ns/op** |
| zlip decode            |      8,551 ns/op |        8,950 ns/op   |
| zlip encode/decode     |     42,813 ns/op |     **16,150 ns/op** |
| identity encode/decode |      610.9 ns/op |        624.8 ns/op   |

**Comment by srebhan** [comment]
@zekena2 now you know why I wanted to have the benchmarks... ;-P

**Review by powersj — APPROVED** [review]
Works for me! Thank you for actually checking the benchmark values

**Comment by zekena2** [comment]
I think klauspost/compress/gzip version is still better than stdlib.
I ran `go test -bench . -v`

goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/internal
cpu: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz

| compression              | stdlib | klauspost |
| :---------------- | :------: | ----: |
| gzip encode        |   21079 ns/op   | **4195 ns/op** |
| gzip decode           |   1434 ns/op   | 1540 ns/op |
| gzip encode/decode    |  22248 ns/op   | **5687 ns/op** |

**Comment by zekena2** [comment]
I've access to graviton instances m6g and m7g and ran the tests there as well, as I remember @powersj asked to run and compare the performance on arm64. m6g is similar to the tests [here](https://github.com/influxdata/telegraf/pull/13316#issuecomment-1581455466) except for big Decode it's 5603364 ns/op.
M7g is so fast in almost everything.

| compression            | m7g           |         new benchmark [here](https://github.com/influxdata/telegraf/pull/13316#issuecomment-1581455466)          |
|------------------------|------------------|----------------------|
| gzip encode            |     **10,341 ns/op** |      16,329 ns/op    |
| gzip decode            |      1,367 ns/op |       1,635 ns/op    |
| gzip encode/decode     |     **11,867 ns/op** |      18,076 ns/op    |
| gzip big encode        | **2,301,083 ns/op** |  3,794,025 ns/op |
| gzip big decode        |  4,193,419 ns/op |    4,469,810 ns/op   |
| gzip big encode/decode | **6,611,384 ns/op** | 12,543,741 ns/op |
| zlip encode            |     3,908 ns/op |      3,348 ns/op |
| zlip decode            |      **5,703 ns/op** |        8,950 ns/op   |
| zlip encode/decode     |     **10,357 ns/op** |     16,150 ns/op |
| identity encode/decode |      616.4 ns/op |        624.8 ns/op   |

and stlib is still slower on both graviton instances as well and using klauspost is faster with similar results as my comment above.

File: internal/content_coding.go [hunk 9]
```diff
@@ -133,19 +149,38 @@ type ContentEncoder interface {
...
+	}
+
 	var buf bytes.Buffer
+	w, err := pgzip.NewWriterLevel(&buf, cfg.level)
 	return &GzipEncoder{
-		writer: gzip.NewWriter(&buf),
-		buf:    &buf,
-	}
+		pwriter: w,
+		writer:  gzip.NewWriter(&buf),
```
Comment (zekena2) [inline internal/content_coding.go:167]: Should this be NewWriterLevel and initialized before return?

Comment (srebhan) [reply internal/content_coding.go:167]: You are right. I would think that now we have the level we should use it...

**Comment by srebhan** [comment]
@zekena2 can you please fix the `docs/LICENSE_OF_DEPENDENCIES.md` file and run `make tidy`? For the level that is a "nice-to-have" but would probably also qualify for a bug-fix PR... Please remember that I would love to have this in v1.27 on Monday...

File: internal/content_coding.go (hunk 9 above)
Comment (zekena2) [inline internal/content_coding.go:167]:
```suggestion
        pw, err := pgzip.NewWriterLevel(&buf, cfg.level)
        if err != nil {
                return nil, err
        }
        w, err := gzip.NewWriterLevel(&buf, cfg.level)
        return &GzipEncoder{
                pwriter: pw,
                writer:  w,
```

@srebhan WDYT to change it to be like this?

Comment (srebhan) [reply internal/content_coding.go]: Agree.

**Review by srebhan — APPROVED** [review]

**Comment by srebhan** [comment]
@powersj can you give it a final review so we are not missing anything here?!?

**Comment by zekena2** [comment]
I added the latest benchmarks between klauspost libraries and stdlib.

**Review by Hipska — APPROVED** [review]

**Review by powersj — APPROVED** [review]
